### PR TITLE
libra-node: periodically dump interesting state

### PR DESCRIPTION
**Pull request by bmwill**
Periodically dump some interesting state:

    * Config and arguments used to start up the process
    * latest ledger version and root hash at that version as well as the
      chain_id

Fixes: #5634

### CODE REVIEW

File: libra-node/src/lib.rs [hunk 1]
```diff
@@ -433,6 +473,12 @@ pub fn setup_environment(node_config: &NodeConfig, logger: Option<Arc<Logger>>)
         debug!("Consensus started in {} ms", instant.elapsed().as_millis());
     }
 
+    // Spawn a task which will periodically dumb some interesting state
```
Comment (coryhill) [inline libra-node/src/lib.rs:476]: > dumb

Freudian slip?

File: libra-node/src/lib.rs [hunk 2]
```diff
@@ -220,6 +220,46 @@ fn setup_debug_interface(config: &NodeConfig, logger: Option<Arc<Logger>>) -> No
     NodeDebugService::new(addr, logger)
 }
 
+async fn periodic_state_dumb(node_config: NodeConfig, db: DbReaderWriter) {
```
Comment (gregnazario) [inline libra-node/src/lib.rs:223]: dump?

Maybe just call it periodic_state :)

File: libra-node/src/lib.rs [hunk 3]
```diff
@@ -220,6 +220,46 @@ fn setup_debug_interface(config: &NodeConfig, logger: Option<Arc<Logger>>) -> No
...
+    use futures::stream::StreamExt;
+
+    let args: Vec<String> = ::std::env::args().collect();
+
+    // Once an hour
+    let mut config_interval = tokio::time::interval(std::time::Duration::from_secs(60 * 60)).fuse();
```
Comment (gregnazario) [inline libra-node/src/lib.rs:229]: I wonder does the compiler precompile the 60 * 60 in this state, and store 3600?

File: libra-node/src/lib.rs [hunk 4]
```diff
@@ -220,6 +220,46 @@ fn setup_debug_interface(config: &NodeConfig, logger: Option<Arc<Logger>>) -> No
...
+    // Once a minute
+    let mut version_interval = tokio::time::interval(std::time::Duration::from_secs(60)).fuse();
```
Comment (bmwill) [inline libra-node/src/lib.rs:231]: @coryhill thoughts on the timing of these state dumps?

File: libra-node/src/lib.rs [hunk 5]
```diff
@@ -220,6 +220,46 @@ fn setup_debug_interface(config: &NodeConfig, logger: Option<Arc<Logger>>) -> No
...
+
+                info!(
+                    chain_id = chain_id,
+                    latest_ledger_verion = latest_ledger_verion,
+                    root_hash = root_hash,
+                    "latest ledger version and its coorisponding root hash"
+                );
+            }
+        }
+    }
```
Comment (gregnazario) [inline libra-node/src/lib.rs:260]: It's funny, this is what we probably want sometimes for the `sample!` macro

File: libra-node/src/lib.rs [hunk 6]
```diff
@@ -220,6 +220,46 @@ fn setup_debug_interface(config: &NodeConfig, logger: Option<Arc<Logger>>) -> No
...
+                };
+
+                let latest_ledger_verion = ledger_info.ledger_info().version();
+                let root_hash = ledger_info.ledger_info().transaction_accumulator_hash();
+
+                info!(
+                    chain_id = chain_id,
+                    latest_ledger_verion = latest_ledger_verion,
+                    root_hash = root_hash,
+                    "latest ledger version and its coorisponding root hash"
```
Comment (gregnazario) [inline libra-node/src/lib.rs:256]: *corresponding

File: libra-node/src/lib.rs (hunk 1 above)
Comment (gregnazario) [reply libra-node/src/lib.rs]: double dumb over dump :)

**Review by gregnazario — APPROVED** [review]
Either rename `dumb` or remove it just to `periodic_state`

File: libra-node/src/lib.rs (hunk 2 above)
Comment (bmwill) [reply libra-node/src/lib.rs:223]: dang i messed up the actual function name too?

File: libra-node/src/lib.rs (hunk 3 above)
Comment (bmwill) [reply libra-node/src/lib.rs:229]: yes it would

File: libra-node/src/lib.rs (hunk 4 above)
Comment (coryhill) [reply libra-node/src/lib.rs:231]: Sounds perfect to me

**Review by davidiw — APPROVED** [review]

File: libra-node/src/lib.rs [hunk 7]
```diff
@@ -220,6 +220,46 @@ fn setup_debug_interface(config: &NodeConfig, logger: Option<Arc<Logger>>) -> No
...
+    use futures::stream::StreamExt;
+
+    let args: Vec<String> = ::std::env::args().collect();
+
+    // Once an hour
+    let mut config_interval = tokio::time::interval(std::time::Duration::from_secs(60 * 60)).fuse();
+    // Once a minute
+    let mut version_interval = tokio::time::interval(std::time::Duration::from_secs(60)).fuse();
+
+    info!("periodic_state_dumb task started");
```
Comment (coryhill) [inline libra-node/src/lib.rs:233]: At least the "dumb" spelling is consistent :D

Comment (bmwill) [reply libra-node/src/lib.rs:233]: another time! jeez I must really not have liked writing this code

**Comment by dariorussi** [comment]
I really would not push this now and wait to take it as a hotfix. 
That is my view obviously ;)

**Comment by bmwill** [comment]
@dariorussi so you think this PR is as dumb as I do? ;)

**Comment by dariorussi** [comment]
I think last second (literally) with a 2 hours conversation in slack should never make the bar and could be taken as hotfix.
That is what I think and I don't even want to look at the PR, it makes no sense.
But as I said in another thread do as you like

**Comment by dariorussi** [comment]
> I think last second PR (literally) with a 2 hours conversation in slack should never make the bar and could be taken as hotfix.
> That is what I think and I don't even want to look at the PR, it makes no sense.
> But as I said in another thread do as you like

it turns out everybody else in the gatekeeper group wants this in 😄 so I have been overruled 'drammatically" - lol. Go for it @bmwill

**Comment by dariorussi** [comment]
> > I think last second PR (literally) with a 2 hours conversation in slack should never make the bar and could be taken as hotfix.
> > That is what I think and I don't even want to look at the PR, it makes no sense.
> > But as I said in another thread do as you like
> 
> it turns out everybody else in the gatekeeper group wants this in 😄 so I have been overruled 'drammatically" - lol. Go for it @bmwill


and you did already anyway ;)

**Pull request merged**